library/alloc/src/collections/linked_list.rs

```diff
@@ -193,13 +204,14 @@ impl<T> LinkedList<T> {
             }
 
             self.len -= 1;
-            node
+            *node
         })
     }
 
     /// Adds the given node to the back of the list.
     #[inline]
-    fn push_back_node(&mut self, mut node: Box<Node<T>>) {
+    fn push_back_node(&mut self, node: Node<T>) {
+        let mut node = Box::new_in(node, &self.alloc);
```

Comment:
I'm worried that changes like this may cause extra stack space etc to be used for moving the Node value around, since there's more places where we're passing it prior to sticking it in the Box. I'm not sure it's worth blocking over, but wanted to raise a note of caution.

Reply:
I've thought of some ways to avoid moving the node onto the stack. Passing `&mut self, mut node: Box<Node<T>, &A>` doesn't work because we can't mutably borrow the whole list while the box is borrowing the list's allocator. Unless we want to add a `Clone` bound on the allocator, we'll have to change one of the arguments.

- The method doesn't mutate the allocator, so we could change the signature to take `head, tail, len` (or some struct containing them) instead of the whole list struct.
- Alternatively, we could pass in a `NonNull<Node<T>>` instead of a `Box`.

Reply:
@Mark-Simulacrum would the following signature be acceptable?
`unsafe fn push_back_node(&mut self, node: Unique<Node<T>>)`

I haven't found a way for this method to safely take a `Box<Node<T>, &A>`, since:
- The box can't hold `&self.alloc` when the function already takes `&mut self`
- The box might use a different instance of `A` with a different memory pool

Reply:
It seems to me that the caller of this function can call Box::new_in, reaching into the LinkedList for the actual allocator, and then pass in `node: Box<Node<T>, A>` just fine, right? Otherwise none of this would actually work.

Reply:
Or I guess `NonNull<Node<T>>` would have to be it. In any case, I'm inclined to not block on this particular function - we can figure that out at a later point. I think focusing on the other comments around changes in behavior is more important.

Reply:
> Unless we want to add a Clone bound on the allocator,

FWIW, BTreeMap has such a bound to avoid `&A` allocators -- those take up space in the Box even for the common case where A is a ZST. (I still think we should [remove the blanked impl](https://github.com/rust-lang/rust/issues/98232) of `&Allocator` to avoid people accidentally blowing up their boxes to twice their size... though in this case it's only temporaries within a function so it's not so bad.)